models/state.go

```diff
@@ -0,0 +1,56 @@
+package models
+
+// State - iota enum of possible image states
+type State int
+
+// Possible values for a State of an image. It can only be one of the following:
+const (
+	StateCreated State = iota
+	StateUploaded
+	StatePublishing
+	StatePublished
+	StateDeleted
+)
+
+var stateValues = []string{"created", "uploaded", "publishing", "published", "deleted"}
+
+// String returns the string representation of a state
+func (s State) String() string {
+	return stateValues[s]
+}
+
+// TransitionAllowed returns true only if the transition from the current state and the provided targetState is allowed
+func (s State) TransitionAllowed(target State) bool {
+	switch s {
+	case StateCreated:
+		switch target {
+		case StateCreated, StateUploaded, StateDeleted:
```

Comment:
Was `StateCreated ` supposed to be left in this list?

Reply:
good point - I thought I left the self-transitions in all states, but apparently I only did it in this one. At the moment this is not used, but I created it as part of the model for validation (It wasn't an original requirement, but I think it will be useful to use this function instead of ad-hoc checks). I think we should allow all self-transitions (e.g. update image keeping the same state)